java/ql/src/experimental/Security/CWE/CWE-470/UnsafeReflectionLib.qll

```diff
@@ -0,0 +1,94 @@
+import java
+import DataFlow
+import semmle.code.java.Reflection
+import semmle.code.java.dataflow.DataFlow
+import semmle.code.java.dataflow.DataFlow3
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.dataflow.TaintTracking2
+
+/**
+ * A call to a Java standard library method which constructs or returns a `Class<T>` from a `String`.
+ * e.g `Class.forName(...)` or `ClassLoader.loadClass(...)`
+ */
+class ReflectiveClassIdentifierMethodAccessCall extends MethodAccess {
+  ReflectiveClassIdentifierMethodAccessCall() {
+    this instanceof ReflectiveClassIdentifierMethodAccess
+  }
+}
+
+/**
+ * Unsafe reflection sink.
+ * e.g `Constructor.newInstance(...)` or `Method.invoke(...)` or `Class.newInstance()`.
+ */
+class UnsafeReflectionSink extends DataFlow::ExprNode {
+  UnsafeReflectionSink() {
+    exists(MethodAccess ma |
+      (
+        ma.getMethod().hasQualifiedName("java.lang.reflect", "Constructor<>", "newInstance")
+        or
+        ma.getMethod().hasQualifiedName("java.lang.reflect", "Method", "invoke")
+      ) and
+      ma.getQualifier() = this.asExpr() and
+      exists(ReflectionArgsConfig rac | rac.hasFlowToExpr(ma.getAnArgument()))
+    )
+  }
+}
+
+/** Taint-tracking configuration tracing flow from remote sources to specifying the initialization parameters to the constructor or method. */
+class ReflectionArgsConfig extends TaintTracking2::Configuration {
```

Comment:
This seems to be the same as the main config but with different additional taint steps. Why are there two configurations?

Reply:
This configuration is to ensure that the construction method parameter value or method parameter value can be controlled by the user remote input.